vulcan/lib/client/jsx/api_types.ts

```diff
@@ -20,7 +20,8 @@ export const TYPE = {
   NESTED_SELECT_AUTOCOMPLETE: 'nested-select-autocomplete',
   MULTIPLE_MULTISELECT_STRING_ALL: 'multiple-multiselect-string-all',
   MULTIPLE_STRING: 'multiple-string',
-  SINGLE_DROPDOWN_MULTICHECKBOX: 'single-dropdown-multicheckbox'
+  SINGLE_DROPDOWN_MULTICHECKBOX: 'single-dropdown-multicheckbox',
+  SCATTER_PLOTLY: 'scatter-plotly'
```

Comment:
Just, to clarify, by the way.  @dtm2451 @coleshaw Is this component a ui-query, or a ui-output?

That is to say, is the intention of this component that it show the result of a workflow, or is the intention of the component to pass results onto to further steps in a workflow?

I suspect that we need to transition this from being an input type to being an output type.

Reply:
I think the component itself looks like an input -- perhaps the names could be changed to make it more clear? Like `plotly-scatter-config`? It does kind of sound like an output widget with the current name.

Reply:
Hmmm, interesting.  I didn't think of there being any distinction like that.

It could go either way perhaps?  But right now, it's written around being a ui-input.  the output of the step is passed to a python script which does some extra bits before calling the scatter_plot function that's in archimedes.

Reply:
If the output is passed on, it's a ui-query.
The reason why you would want a ui-output and not a ui-query is if you intentionally did not want to _promise the interface of the output_.  For instance, a component which may change its internal data format quite frequently as features are added will be hard to maintain as a ui-query, since any other workflow that depends on that component will instantly break when value type changes.

As a rule, once we use a ui-query in a workflow, _we may never again change its value type_, since we must remain compatible with all existing workflows.

If we think that a component is oft going to change its value, we're talking about a ui-output that makes no promises to connect other steps in other workflows.

Reply:
Hmmm.  I'd thought about the issue of the component <-> script connectivity a bit before, but certainly not with the full scale of this maintainability issue in mind.  I think we do want this to be a ui-output then.  

That raises 2 issues:
1. The plotting code is currently in python, though we could probably migrate it to React-python.
2. Color-data retrieval currently happens after the step where I was going to plug this into UMAP.  I think we want to keep that retrieval behind a ui-input because querying all options of data may get expensive for large data with many options & possibly also buggy if there's missing data.

For 2, we can leave the color-by selection method as-was before this PR, then have plot adjustments be a separate (often optional) ui bit.

For 1... I imagine it would be better to plan on making future visualizations directly in React than to create a back-end path where a ui-output can require a paired python script be run after inputs are commited.

Reply:
Or, alternatively, perhaps we can build a `stable bridge` between react <-> plotly?

So instead of a bunch of on off scripts + one off configurations, there is a single plotly script that accepts a full plotly configuration?  And therefore, all python side plotly interactions involve just one script and one type of input (but the ui-queries can be of various kinds, each simply always outputing the bridge config type?)

Reply:
Might that basically be the same as the Plotly.JS JSON `layout` (and possibly `config`) payloads that get sent from the python to the `Plot.tsx` output component? So any "plot config" UI inputs would write to the Plotly format, regardless of what kind of plot they are configuring? Hopefully that made sense...

So the interface would be the same [as outlined in this Plotly.js documentation](https://plotly.com/javascript/reference/layout/).

Reply:
Yeah, that's more or less what I'm thinking.  Especially if the python side isn't really "doing" much beyond that.

Reply:
`stable bridge` sounds like what I was already planning for the archimedes plotting folder... A single function per plot type that parses our plot setups into how those actually need to provide those to plotly.

Reply:
It's already the case that the script following this widget in UMAP just adds some set-within-the-workflow inputs (behind the `hide` option of this widget) to the payload of inputs for the `import scatter_plot from archimedes.functions.plotting` call.  Then it makes the call.  Then it `output_json`s that return ("to" the ui-output step that renders the plot).  So, the code change on the workflow side would be a reorganization rather than an algorithmic restructure :+1: